### PR TITLE
Inner loop networking article

### LINKED ISSUE #232: .NET Aspire networking documentation
We need a section in the TOC that covers off the interaction between apps and containers running locally in Aspire and the Traefik proxies that DCP manages. We don't necessarily need to mention DCP or traefik as they are mostly implementation details from an Aspire perspective but we need to make it clear that when a service binding is added either implicitly (via launch profiles) or explicitly via a manually added service binding that Aspire will launch a Traefik proxy on that port, and the executable, project or container will be assigned another port.

In the case of Projects we automatically inject the appropriate ASPNETCORE_URLS environment variables. For containers we randomly assign the ports using the magic of Docker. For executables it requires some integration where we will optionally inject an environment variable with the port assigned for each service binding. This means that the executable will need to use that variable to figure out what port to bind to locally in order to avoid conflicts.

Networking isn't an inner loop only consideration. We probably need pages to cover various deployment scenarios and considerations and limitations of each target environment. There is not always a 1:1 mapping between what we do locally, and what gets deployed to the target environment.

---
[Associated WorkItem - 198655](https://dev.azure.com/msft-skilling/Content/_workitems/edit/198655)

### PR DESCRIPTION
**Pull request by IEvangelist**
## Summary

Inner loop networking article

Fixes #232

Hey, @davidfowl and @mitchdenny - here's a draft PR for the inner loop networking content. Please let me know what you think.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/fundamentals/networking-overview.md](https://github.com/dotnet/docs-aspire/blob/2659809815a33b55ba1f5ea56bd839cea3516c85/docs/fundamentals/networking-overview.md) | [.NET Aspire inner-loop networking overview](https://review.learn.microsoft.com/en-us/dotnet/aspire/fundamentals/networking-overview?branch=pr-en-us-279) |

### CODE REVIEW

File: docs/fundamentals/networking-overview.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
+---
+title: .NET Aspire inner loop networking overview
+description: Learn how .NET Aspire handles networking and service bindings, and how you can use them in your app code.
+ms.date: 01/08/2024
+---
```
Comment (gewarren) [inline docs/fundamentals/networking-overview.md:5]:
```suggestion
ms.topic: overview
---
```

File: docs/fundamentals/networking-overview.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+# .NET Aspire inner loop networking overview
```
Comment (gewarren) [inline docs/fundamentals/networking-overview.md:7]:
```suggestion
# .NET Aspire inner-loop networking overview
```

File: docs/fundamentals/networking-overview.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+One of the advantages of developing with .NET Aspire is that it enables you to develop, test, and debug cloud-native apps locally. Inner loop networking is a key aspect of .NET Aspire, that allows your apps to communicate with each other in your development environment. In this article, you learn how .NET Aspire handles various networking scenarios, with proxies, service bindings, endpoint configurations, and launch profiles.
```
Comment (gewarren) [inline docs/fundamentals/networking-overview.md:9]:
```suggestion
One of the advantages of developing with .NET Aspire is that it enables you to develop, test, and debug cloud-native apps locally. Inner-loop networking is a key aspect of .NET Aspire that allows your apps to communicate with each other in your development environment. In this article, you learn how .NET Aspire handles various networking scenarios with proxies, service bindings, endpoint configurations, and launch profiles.
```

File: docs/fundamentals/networking-overview.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+- **Launch profiles**: Launch profiles are configuration files that specify how to run your app locally. You can use launch profiles (such as the _launchSettings.json_ file) to define the service bindings, environment variables, and launch settings for your app.
+- **Service bindings/Endpoint configurations**: Service bindings are the connections between your app and the services it depends on, such as databases, message queues, or APIs. Service bindings provide information such as the service name, host port, scheme, and environment variable. You can add service bindings to your app either implicitly (via launch profiles) or explicitly by calling <xref:Aspire.Hosting.ResourceBuilderExtensions.WithServiceBinding%2A>.
+- **Proxies**: .NET Aspire automatically launches a proxy for each service binding you add to your app, and assigns a port for the proxy to listen on. The proxy then forwards the requests to the port that your app listens on, which may be different from the proxy port. This way, you can avoid port conflicts and access your app and services using consistent and predictable URLs.
```
Comment (gewarren) [inline docs/fundamentals/networking-overview.md:17]:
```suggestion
- **Proxies**: .NET Aspire automatically launches a proxy for each service binding you add to your app, and assigns a port for the proxy to listen on. The proxy then forwards the requests to the port that your app listens on, which might be different from the proxy port. This way, you can avoid port conflicts and access your app and services using consistent and predictable URLs.
```

File: docs/fundamentals/networking-overview.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+## How service bindings work
+
+A service binding in .NET Aspire involves two components: a **service** representing an external resource your app requires (for example, a database, message queue, or API), and a **binding** establishing a connection between your app and the service, providing necessary information.
```
Comment (gewarren) [inline docs/fundamentals/networking-overview.md:21]:
```suggestion
A service binding in .NET Aspire involves two components: a **service** representing an external resource your app requires (for example, a database, message queue, or API), and a **binding** that establishes a connection between your app and the service and provides necessary information.
```

File: docs/fundamentals/networking-overview.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+.NET Aspire supports two service binding types: **implicit**, automatically created based on specified launch profiles defining app behavior in different environments, and **explicit**, manually created using <xref:Aspire.Hosting.ResourceBuilderExtensions.WithServiceBinding%2A>.
+
+Upon creating a binding, whether implicit or explicit, .NET Aspire launches a lightweight reverse proxy on a specified port, handling routing and load balancing for requests from your app to the service. The proxy is a .NET Aspire implementation detail, requiring no configuration or management concern.
+
+To help visualize how service bindings work, consider the .NET Aspire starter templates inner loop networking diagram:
```
Comment (gewarren) [inline docs/fundamentals/networking-overview.md:27]:
```suggestion
To help visualize how service bindings work, consider the .NET Aspire starter templates inner-loop networking diagram:
```

File: docs/fundamentals/networking-overview.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+:::image type="content" source="media/networking/networking-proxies-1x.png" lightbox="media/networking/networking-proxies.png" alt-text=".NET Aspire Starter Application template inner loop networking diagram.":::
+
+## Launch profiles
+
+When you call <xref:Aspire.Hosting.ProjectResourceBuilderExtensions.AddProject%2A>, the app host looks for _Properties/launchSettings.json_ to determine the default set of service bindings. The app host selects a specific launch profile using the following rules:
+
+1. An explicit <xref:Aspire.Hosting.ProjectResourceBuilderExtensions.WithLaunchProfile%2A> call on the `IResourceBuilder<ProjectResource>`.
+1. The `DOTNET_LAUNCH_PROFILE` environment variable. For more information, see [.NET environment variables](/dotnet/core/tools/dotnet-environment-variables).
+1. Picking the first launch profile defined in _launchSettings.json_.
```
Comment (gewarren) [inline docs/fundamentals/networking-overview.md:37]:
```suggestion
1. The first launch profile defined in _launchSettings.json_.
```

File: docs/fundamentals/networking-overview.md [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+To specify that the **https** launch profile should be used, call <xref:Aspire.Hosting.ProjectResourceBuilderExtensions.WithLaunchProfile%2A>:
+
+:::code source="snippets/networking/Networking.AppHost/Program.WithLaunchProfile.cs" id="withlaunchprofile":::
+
+This will select the **https** launch profile from _launchSettings.json_. The `applicationUrl` of that launch profile is used to create a service binding for this project. This is the equivalent of:
+
+:::code source="snippets/networking/Networking.AppHost/Program.WithLaunchProfile.cs" id="verbose":::
+
+> [!IMPORTANT]
+> If there's no _launchSettings.json_ (or launch profile) there will be no bindings by default.
```
Comment (gewarren) [inline docs/fundamentals/networking-overview.md:58]:
```suggestion
> If there's no _launchSettings.json_ (or launch profile), there are no bindings by default.
```

File: docs/fundamentals/networking-overview.md [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+The preceding diagram depicts the following:
+
+- A web browser as an entry point to the app.
+- A host port of 5066.
+- The frontend proxy sitting between the web browser and the frontend service replicas, listening on port 5066.
+- The `frontend_0` frontend service replica listening on the randomly assigned port 65001.
+- The `frontend_1` frontend service replica listening on the randomly assigned port 65002.
+
+Without the call to `WithReplicas`, there's only one frontend service. The proxy will still be listening on port 5066, but the frontend service will be listening on a random port:
```
Comment (gewarren) [inline docs/fundamentals/networking-overview.md:78]:
```suggestion
Without the call to `WithReplicas`, there's only one frontend service. The proxy still listens on port 5066, but the frontend service listens on a random port:
```

File: docs/fundamentals/networking-overview.md [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+- A web browser as an entry point to the app.
+- A host port of 5066.
+- The frontend proxy sitting between the web browser and the frontend service, listening on port 5066.
+- The frontend service listening on random port of 65001.
+
+The underlying service is fed this port via `ASPNETCORE_URLS` for project resources. Other resources access to this port by specifying an environment variable on the service binding:
+
+:::code source="snippets/networking/Networking.AppHost/Program.EnvVarPort.cs" id="envvarport":::
+
+The above code will make the random port available in the `PORT` environment variable. The app uses this to listen to incoming connections from the proxy. Consider the following diagram:
```
Comment (gewarren) [inline docs/fundamentals/networking-overview.md:100]:
```suggestion
The previous code makes the random port available in the `PORT` environment variable. The app uses this port to listen to incoming connections from the proxy. Consider the following diagram:
```

File: docs/fundamentals/networking-overview.md [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+The preceding diagram depicts the following:
+
+- A web browser as an entry point to the app.
+- A random host port of 65000.
+- The frontend proxy sitting between the web browser and the frontend service, listening on port 65000.
+- The frontend service listening on a random port of 65001.
+
+## Container ports
+
+When you add a container resource, .NET Aspire will automatically assign a random port to the container. To specify a container port, configure the container resource with the desired port:
```
Comment (gewarren) [inline docs/fundamentals/networking-overview.md:130]:
```suggestion
When you add a container resource, .NET Aspire automatically assigns a random port to the container. To specify a container port, configure the container resource with the desired port:
```

**Review by gewarren — APPROVED** [review]

**Comment by davidfowl** [comment]
We're missing the piece about `WithReference` and how it uses the proxy address

**Comment by IEvangelist** [comment]
> We're missing the piece about `WithReference` and how it uses the proxy address

Is that not covered in the opening paragraph of the [Ports and proxies section](https://review.learn.microsoft.com/en-us/dotnet/aspire/fundamentals/networking-overview?branch=pr-en-us-279#ports-and-proxies)?

**Comment by davidfowl** [comment]
👍🏾

**Pull request merged**